jni/src/common/Foreign/JNI/Internal/BackgroundWorker.hs

```diff
@@ -55,15 +68,21 @@ stop worker =
   where
     submitStopAtEndOfBatch = do
       let task = throwIO StopWorkerException
-      atomicModifyIORef (nextBatch worker) $ \tasks -> (tasks >> task, ())
+      atomicModifyIORef (nextBatch worker) $ \(n, tasks) -> ((n, tasks >> task), ())
       void $ tryPutMVar (wakeup worker) ()
 
 -- | Submits a task to the background worker.
 --
 -- Any exception thrown in the given task will stop the
 -- background thread.
 --
+-- If the job queue is currently full, block until it isn't
+--
 submitTask :: BackgroundWorker -> IO () -> IO ()
 submitTask worker task = do
-  atomicModifyIORef (nextBatch worker) $ \tasks -> (task >> tasks, ())
+  maxSize <- readIORef $ queueSize worker
+  flip whileM_ (return ()) $ atomicModifyIORef (nextBatch worker) $ \(n, tasks) ->
+    if n >= maxSize
+    then ((n, tasks), True)
+    else ((n+1, task >> tasks), False)
```

Comment:
This is a tough one. I see two paths.

One is to use a [Broadcast](https://hackage.haskell.org/package/concurrent-extra-0.7.0.12/docs/Control-Concurrent-Broadcast.html), this probably will add some overhead, although we would pay for it only when the queue is full.

The other option is to implement the broadcast with [STM](http://hackage.haskell.org/package/stm).

We could try to replace the IORef with a TVar as well, though this would require some benchmarking of the non-full case.

Reply:
I found these benchmarks: https://codeburst.io/the-haskell-concurrency-primitive-shootout-538c21993f1c
After running them locally, it is looking to me like we should go with a couple of TVars to hold the next batch and to signal that the queue has emptied.